src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -17,6 +17,37 @@ int MPIR_Typerep_pack(const void *inbuf, MPI_Aint incount, MPI_Datatype datatype
     int mpi_errno = MPI_SUCCESS;
     int rc;
 
+    if (incount == 0) {
+        *actual_pack_bytes = 0;
+        goto fn_exit;
+    }
+
+    MPIR_Assert(datatype != MPI_DATATYPE_NULL);
+
+    /* fast-path for contig CPU buffers */
+    bool is_contig;
+    MPIR_Datatype_is_contig(datatype, &is_contig);
+    if (is_contig) {
```

Comment:
Can we get is_contig and true_lb at the same time? I see two pointer dereferences here if the datatype is not built-in.

Reply:
Hmm?  First of all, there's no dereference for builtin types.  Second, these are macros, so it'd essentially be the same as getting both at the same time.

Reply:
One potential optimization would be to do this only for builtin types, rather than contiguous types.  That way, we can get rid of the pointer dereference for size too.  That'll obviously hurt contiguous derived types, though.

Reply:
I was concerning the pointer dereferences for non-built-in continuous datatypes. Why not get is_contig and true_lb at the same time instead of calling each MACRO separately?

Reply:
You are right that we can get size both at the same time. For example, using `MPIDI_Datatype_check_contig_size_lb` before the if check.

If it is not contiguous, we don't care the extra store instructions (for size and true_lb) anyway.

Reply:
Oh, I just notice that `MPIDI_Datatype_check_contig_size_lb` is ch4 internal. Why do we have two sets of datatype attribute functions, some in MPIR but some others in ch4?

Reply:
@minsii  That's a `ch4` macro.  We cannot use it in `typerep`.

Reply:
Also, my point is that, with macros it'd be equivalent.  The compiler will merge the code segments for us.

Reply:
I am not sure if the compiler can detect that the three datatype pointer dereferences are the same. I noticed that the macro is in ch4. But why can't we move it to MPIR if it is useful here?

Reply:
They are not the same.  That's my point.  Which pointer dereferences are the same that you are hoping the compiler can merge if we used a single macro?

Reply:
For non-builtin datatypes, I see the is_contig, size, and true_lb attributes are obtained from the following code in your current version:
```
#define MPIR_Datatype_is_contig(dtype_, is_contig_)                            \
        ...
        else {                                                                 \
            MPIR_Datatype *dtp_ = NULL;                                        \
            MPIR_Datatype_get_ptr((dtype_), dtp_);                             \
            MPIR_Assert(dtp_ != NULL);                                         \
            *(is_contig_) = dtp_->is_contig;                                   \


#define MPIR_Datatype_get_size_macro(a,size_) do {                     \
        ...
        case HANDLE_KIND_DIRECT:                                        \
            MPIR_Assert(HANDLE_INDEX(a) < MPIR_DATATYPE_PREALLOC);      \
            ptr = MPIR_Datatype_direct+HANDLE_INDEX(a);                 \
            size_ = ((MPIR_Datatype *) ptr)->size;                      \
            break;                                                      \

#define MPIR_Datatype_get_true_lb(dtype_, true_lb_)                            \
        ...
        else {                                                                 \
            MPIR_Datatype *dtp_ = NULL;                                        \
            MPIR_Datatype_get_ptr((dtype_), dtp_);                             \
            *(true_lb_) = dtp_->true_lb;                                       \
```

Can they be combined similar to `MPIDI_Datatype_check_contig_size_lb `? Please see code capture below. Then the `MPIR_Datatype_get_ptr` will be called only once.
```
            MPIR_Datatype *dt_ptr_;                             \
            MPIR_Datatype_get_ptr((datatype_), (dt_ptr_));      \
            (dt_contig_out_) = (dt_ptr_)->is_contig;        \
            (data_sz_out_)   = (size_t)(count_) *           \
                (dt_ptr_)->size;                            \
            (dt_true_lb_)    = (dt_ptr_)->true_lb;          \
```

Reply:
Your final "optimized" example still has the three pointer deferences:

```
            (dt_contig_out_) = (dt_ptr_)->is_contig;        \
            (data_sz_out_)   = (size_t)(count_) *           \
                (dt_ptr_)->size;                            \
            (dt_true_lb_)    = (dt_ptr_)->true_lb;          \
```

Where are you saving pointer dereferences?

Reply:
Oops, sorry. Wrong term here. I meant the computation for getting the pointer of the datatype. I.e.,
```
        case HANDLE_KIND_DIRECT:                                        \
            ptr=MPIR_##kind##_direct+HANDLE_INDEX(a);                   \
```

Reply:
I see.  That's not a pointer dereference, though.  That's just an arithmetic operation.  I guess it is possible to save that too, but that really looks like a stretch at this point.  :-)

I'll workaround it in my code, because I don't want to argue about it anymore.  :-)

Reply:
Is this new version better?

Reply:
Yikes, hold off on the review.  I tried to combine the changes to a single function, but screwed up which buffer the true_lb should apply to.